### PR TITLE
Improving SAL message strings

### PR DESCRIPTION
**Pull request by eddynaka**

### CODE REVIEW

**Comment by michaelcfanning** [comment]
>       "Message": "'{0:scanTarget}' contains a call to the SAL v1 '{1:obsoleteApi}' annotation and should be removed as it is obsolete or redundant in SAL v2.",

'{0:scanTarget}' contains a use of the obsolete or redundant '{obsoleteAnnotation}' annotation that should be removed.

---
Refers to: Src/Plugins/SalModernization/SEC105.UpdateSalToCurrentVersion.json:8 in 2881e4f. [](commit_id = 2881e4fbb9763450a291dcbb4ea394f90b3c8032, deletion_comment = False)

File: Src/Plugins/SalModernization/SalModernization.SharedStrings.txt [hunk 1]
```diff
@@ -21,636 +21,632 @@
...
-$SEC105/626.ReturnValuePostAccessSAWrite=\[returnvalue:Post\(Access=SA_Write\)\]
-$SEC105/627.ReturnValuePostAccessSAReadWriteSARead=\[returnvalue:Post\(Access=\(SA_ReadWrite\s{0,4}|\s{0,4}SA_Read\)\)\]
-$SEC105/628.ReturnValuePostValidSAMaybeDeref1=\[returnvalue:Post\(Valid=SA_Maybe,\s{0,4}Deref=1\)\]
-$SEC105/629.ReturnValueSAPostDeref1NullSAYesNullTerminatedSANo=\[returnvalue:\s{0,4}SA_Post\(\s{0,4}Deref=\(1\),\s{0,4}Null=SA_Yes,\s{0,4}NullTerminated=SA_No\s{0,4}\)\]
-$SEC105/630.ReturnValuePostAccessSAMaybe=\[returnvalue:Post\(Access=SA_Maybe\)\]
-$SEC105/631.ReturnValuePostAccessSANoAccess=\[returnvalue:Post\(Access=SA_NoAccess\)\]
-$SEC105/632.ReturnValuePostAccessSANoAccessSANoAccess=\[returnvalue:Post\(Access=\(SA_NoAccess\s{0,4}|\s{0,4}SA_NoAccess\)\)\]
-$SEC105/633.ReturnValuePostDeref1ValidSAYes=\[returnvalue:Post\(Deref='%1',\s{0,4}Valid=SA_Yes\)\]
\ No newline at end of file
+$SEC105/001/001.__allocator=[^\w_]__allocator[^\w_]
```
Comment (michaelcfanning) [inline Src/Plugins/SalModernization/SalModernization.SharedStrings.txt:24]: >$SEC105/001/001 [](http://example.com/codeflow?start=0&length=15)

It isn't right to have two levels of number in an opaque id, no other analysis pattern does this and it's hard to read. Instead, drop the final id.

So this s/be $SEC105/001.allocator

**Comment by michaelcfanning** [comment]
>             "fix": {

This name should be descriptive of the fix. The reason is that we may need to specify this label when auto-fixing from the command-line. Maybe use 'deleteAnnotation'?

---
Refers to: Src/Plugins/SalModernization/SEC105.UpdateSalToCurrentVersion.json:17 in 2881e4f. [](commit_id = 2881e4fbb9763450a291dcbb4ea394f90b3c8032, deletion_comment = False)

File: Src/Plugins/Tests.SalModernization/TestData/UpdateSalToCurrentVersion/ExpectedOutputs/SEC105_001.AnalysisAssume.sarif [hunk 2]
```diff
@@ -14,80 +14,48 @@
...
-              "fullDescription": {
-                "text": "Update SAL to current version"
-              },
-              "messageStrings": {
-                "NotApplicable_InvalidMetadata": {
-                  "text": "'{0}' was not evaluated for check '{1}' because the analysis is not relevant for the following reason: {2}."
-                },
-                "Default": {
-                  "text": "'{0}' contains incorrect or out-dated use of SAL."
+                  "text": "'{0}' contains a call to the SAL v1 '{1}' annotation and should be converted manually."
```
Comment (michaelcfanning) [inline Src/Plugins/Tests.SalModernization/TestData/UpdateSalToCurrentVersion/ExpectedOutputs/SEC105_001.AnalysisAssume.sarif:58]: >call [](http://example.com/codeflow?start=44&length=4)

There are no 'calls' to annotations, annotations are applied to code. i.e., they are metadata only, not executable.

'{0}' contains a use of the SAL v1 '{1} annotation that should be converted manually to the correct SAL v2 pattern.

**Comment by michaelcfanning** [comment]
>       "Id": "SEC105/001",

Where is the subid use in your checks? Please go read the SARIF spec about rule ids and perhaps look at the tutorial so that you can form an opinion about what's appropriate here. :)

---
Refers to: Src/Plugins/SalModernization/SEC105.UpdateSalToCurrentVersion.json:6 in 2881e4f. [](commit_id = 2881e4fbb9763450a291dcbb4ea394f90b3c8032, deletion_comment = False)

**Comment by eddynaka** [comment]
>       "Id": "SEC105/001",

I used SEC105/001 or 002 or 003 because i wanted to separate between delete/update/manual.

I didn't want to create SEC105/SEC106/SEC107 for that.

For subid, i thought about that. I didn't use because if we  create one subid per each match expression, if u have  10 different functions to udpate, the result sarif would have 10 equal messages with different messagestrings key.

---
In reply to: [822005250](https://github.com/microsoft/sarif-pattern-matcher/pull/388#issuecomment-822005250) [](http://example.com/codeflow?ancestors=822005250,822005250)

---
Refers to: Src/Plugins/SalModernization/SEC105.UpdateSalToCurrentVersion.json:6 in 2881e4f. [](commit_id = 2881e4fbb9763450a291dcbb4ea394f90b3c8032, deletion_comment = False)

File: Src/Plugins/Tests.SalModernization/TestData/UpdateSalToCurrentVersion/ExpectedOutputs/SEC105_001.AnalysisAssume.sarif (hunk 2 above)
Comment (eddynaka) [reply Src/Plugins/Tests.SalModernization/TestData/UpdateSalToCurrentVersion/ExpectedOutputs/SEC105_001.AnalysisAssume.sarif]: fixed.

File: Src/Plugins/SalModernization/SalModernization.SharedStrings.txt (hunk 1 above)
Comment (eddynaka) [reply Src/Plugins/SalModernization/SalModernization.SharedStrings.txt]: fixed.

**Comment by michaelcfanning** [comment]
>       "Id": "SEC105/001",

I don't think so. A sub-id isn't expressed in a rule definition.

---
In reply to: [822014675](https://github.com/microsoft/sarif-pattern-matcher/pull/388#issuecomment-822014675) [](http://example.com/codeflow?ancestors=822014675,822005250)

---
Refers to: Src/Plugins/SalModernization/SEC105.UpdateSalToCurrentVersion.json:6 in 2881e4f. [](commit_id = 2881e4fbb9763450a291dcbb4ea394f90b3c8032, deletion_comment = False)

File: Src/Plugins/Tests.SalModernization/TestData/UpdateSalToCurrentVersion/ExpectedOutputs/SEC105_001.AnalysisAssume.sarif [hunk 3]
```diff
@@ -14,80 +14,48 @@
           "rules": [
             {
               "id": "SEC105/001",
```
Comment (michaelcfanning) [inline Src/Plugins/Tests.SalModernization/TestData/UpdateSalToCurrentVersion/ExpectedOutputs/SEC105_001.AnalysisAssume.sarif:16]: >001 [](http://example.com/codeflow?start=28&length=3)

You should go back to your bucket rules.

RemoveObsoleteOrRedundantAnnotations
RenameLegacyAnnotationsToCurrentVersion
UpdateAnnotationsToCurrentVersion

Or something.

For each of these, use the name of the annotation as a sub-id. Consult the spec for more information. A sub-id is an additional parameter that is in a rule only, it is not valid in a rule definition.

**Review by michaelcfanning — CHANGES_REQUESTED** [review]
:clock1:

File: Src/Plugins/SalModernization/SalModernization.SharedStrings.txt [hunk 4]
```diff
@@ -21,636 +21,632 @@
...
+$SEC105/579.PostWritableElementsConst1Deref1=\[Post\(WritableElementsConst=%1, Deref=1\)\]
+$SEC105/580.PostDeref1WritableElementsConst1=\[Post\(Deref=1, WritableElementsConst=%1\)\]
+$SEC105/581.PostWritableElements1Deref1=\[Post\(WritableElements="%1", Deref=1\)\]
+$SEC105/582.PostWritableElements1=\[Post\(WritableElements="%1"\)\]
+$SEC105/583.PostDeref1ValidElementsConst1=\[Post\(Deref=1, ValidElementsConst=%1\)\]
+$SEC105/584.PostDeref1Tainted1=\[Post\(Deref=1, Tainted=%1\)\]
+$SEC105/585.PostValidSAMaybeDeref1=\[Post\(Valid=SA_Maybe, Deref=1\)\]
+$SEC105/586.PostAccess1=\[Post\( Access=%1 \)\]
+$SEC105/587.PostDeref1Access1=\[Post\( Deref=1, Access=%1 \)\]
+$SEC105/588.PostDeref1NeedsRelease1=\[Post\( Deref=1, NeedsRelease=%1 \)\]
```
Comment (michaelcfanning) [inline Src/Plugins/SalModernization/SalModernization.SharedStrings.txt:611]: >$SEC105/588 [](http://example.com/codeflow?start=0&length=11)

I don't think this is right. Now every needs to manage hundreds of rules. The help topic for most of these is the same! e.g., 'This annotation is obsolete in v1'. :)

See my other comment. Also, where is the code that is autogenerating all this content, we should check it in.

File: Src/Plugins/Tests.SalModernization/TestData/UpdateSalToCurrentVersion/ExpectedOutputs/SEC105_001.AnalysisAssume.sarif (hunk 3 above)
Comment (eddynaka) [reply Src/Plugins/Tests.SalModernization/TestData/UpdateSalToCurrentVersion/ExpectedOutputs/SEC105_001.AnalysisAssume.sarif:16]: changed that. take a look when u can.

File: Src/Plugins/SalModernization/SalModernization.SharedStrings.txt (hunk 4 above)
Comment (eddynaka) [reply Src/Plugins/SalModernization/SalModernization.SharedStrings.txt]: just did as a private method in the Cli project. just call it pointing to the sal.txt and will make the conversion.

It generates a json and a stringbuilder.

**Comment by michaelcfanning** [comment]
>                 "allocator": {

This is wrong! We should only be using a single message string. :)



---
Refers to: Src/Plugins/Tests.SalModernization/TestData/UpdateSalToCurrentVersion/ExpectedOutputs/SEC105_001.AnalysisAssume.sarif:25 in 82e6849. [](commit_id = 82e6849bc20eff447e557906d8d5440f250027a3, deletion_comment = False)

**Comment by michaelcfanning** [comment]
>                   "text": "'{0}' contains a use of the obsolete or redundant '{1}' annotation and should be replaced for '{2}'."

'{0}' contains a use of the SAL v1 '{1}' annotation that should be replaced with the SAL v2 equivalent '{2}'.

---
Refers to: Src/Plugins/Tests.SalModernization/TestData/UpdateSalToCurrentVersion/ExpectedOutputs/SEC105_001.AnalysisAssume.sarif:90 in 82e6849. [](commit_id = 82e6849bc20eff447e557906d8d5440f250027a3, deletion_comment = False)

**Comment by michaelcfanning** [comment]
>                   "text": "'{0}' contains a use of the obsolete or redundant '{1}' annotation that should be removed."

'{0}' contains a use of the obsolete or redundant SAL v1 '{1}' annotation that should be removed.

---
Refers to: Src/Plugins/Tests.SalModernization/TestData/UpdateSalToCurrentVersion/ExpectedOutputs/SEC105_001.AnalysisAssume.sarif:41 in 82e6849. [](commit_id = 82e6849bc20eff447e557906d8d5440f250027a3, deletion_comment = False)

**Comment by michaelcfanning** [comment]
>                   "text": "'{0}' contains a use of the SAL v1 '{1} annotation that should be converted manually to the correct SAL v2 pattern."

You have a missing apostrophe in your original string.  I also changed 'correct' to 'equivalent'.

'{0}' contains a use of the SAL v1 '{1}' annotation that should be converted manually to the equivalent SAL v2 pattern.

---
Refers to: Src/Plugins/Tests.SalModernization/TestData/UpdateSalToCurrentVersion/ExpectedOutputs/SEC105_001.AnalysisAssume.sarif:1696 in 82e6849. [](commit_id = 82e6849bc20eff447e557906d8d5440f250027a3, deletion_comment = False)

**Comment by michaelcfanning** [comment]
>                 "SAPreNullTerminatedCOMINGNullSANo": {

Some of these names look odd and probably could use a scrub in the original data. I can help make a pass over things.

---
Refers to: Src/Plugins/Tests.SalModernization/TestData/UpdateSalToCurrentVersion/ExpectedOutputs/SEC105_001.AnalysisAssume.sarif:1815 in 82e6849. [](commit_id = 82e6849bc20eff447e557906d8d5440f250027a3, deletion_comment = False)

**Comment by michaelcfanning** [comment]
>                 "allocator": {

There should only be a single string for this rule named 'Default' with a single format message string in it. We don't want to duplicate all this data. This s/be possible with our current model: we have other security checks that share a common string despite defining a sub id. If you need help, let me know. If we have a design issue we need to resolve to make this work, let me know. :)

---
Refers to: Src/Plugins/Tests.SalModernization/TestData/UpdateSalToCurrentVersion/ExpectedOutputs/SEC105_001.AnalysisAssume.sarif:25 in 82e6849. [](commit_id = 82e6849bc20eff447e557906d8d5440f250027a3, deletion_comment = False)

File: Src/Plugins/Tests.SalModernization/TestData/UpdateSalToCurrentVersion/ExpectedOutputs/SEC105_001.AnalysisAssume.sarif [hunk 5]
```diff
@@ -266,7 +238,7 @@
           "fixes": [
             {
               "description": {
-                "text": "Automatic AnalysisAssume"
+                "text": "Replace '__analysis_assume' for '_Analysis_assume_'."
```
Comment (michaelcfanning) [inline Src/Plugins/Tests.SalModernization/TestData/UpdateSalToCurrentVersion/ExpectedOutputs/SEC105_001.AnalysisAssume.sarif:241]: >for [](http://example.com/codeflow?start=53&length=3)

ths should be 'with' not 'for' :)

**Review by michaelcfanning — APPROVED** [review]
:shipit:

**Pull request merged**